File: sdk/python/src/agentmux_sdk/_sse.py
"""Minimal Server-Sent Events parsing shared by the sync and async clients.

The AgentMux invocation stream emits ``event: <type>\\ndata: <json>\\n\\n``
blocks plus ``: keepalive`` comment lines every 15 seconds. Comment lines
(starting with ``:``) must be ignored per the SSE specification.
"""

from __future__ import annotations

import json
from collections.abc import AsyncIterator, Iterator
from typing import Any


class SSEBlockAssembler:
    """Feed decoded text chunks, get complete event blocks back."""

    def __init__(self) -> None:
        self._buffer = ""

    def feed(self, chunk: str) -> list[str]:
        self._buffer += chunk
        blocks: list[str] = []
        while True:
            for separator in ("\r\n\r\n", "\n\n"):
                index = self._buffer.find(separator)
                if index >= 0:
                    blocks.append(self._buffer[:index])
                    self._buffer = self._buffer[index + len(separator):]
                    break
            else:
                return blocks

    def flush(self) -> str:
        remaining, self._buffer = self._buffer, ""
        return remaining


def parse_sse_block(block: str) -> dict[str, Any] | None:
    """Extract the JSON payload of one SSE block; None for comments/empties."""
    data_lines = [
        line[5:].lstrip()
        for line in block.splitlines()
        if line.startswith("data:")
    ]
    if not data_lines:
        return None
    try:
        payload = json.loads("\n".join(data_lines))
    except ValueError:
        return None
    return payload if isinstance(payload, dict) else None


def iter_sse_payloads(chunks: Iterator[str]) -> Iterator[dict[str, Any]]:
    assembler = SSEBlockAssembler()
    for chunk in chunks:
        for block in assembler.feed(chunk):
            payload = parse_sse_block(block)
            if payload is not None:
                yield payload
    tail = assembler.flush()
    if tail.strip():
        payload = parse_sse_block(tail)
        if payload is not None:
            yield payload


async def aiter_sse_payloads(chunks: AsyncIterator[str]) -> AsyncIterator[dict[str, Any]]:
    assembler = SSEBlockAssembler()
    async for chunk in chunks:
        for block in assembler.feed(chunk):
            payload = parse_sse_block(block)
            if payload is not None:
                yield payload
    tail = assembler.flush()
    if tail.strip():
        payload = parse_sse_block(tail)
        if payload is not None:
            yield payload

File: sdk/typescript/src/sse.ts
/**
 * Minimal Server-Sent Events parsing over a ReadableStream.
 *
 * The AgentMux invocation stream emits `event: <type>\ndata: <json>\n\n`
 * blocks plus `: keepalive` comment lines every 15 seconds; comment lines
 * are ignored per the SSE specification.
 */

export function parseSSEBlock(block: string): Record<string, unknown> | null {
  const dataLines = block
    .split(/\r?\n/)
    .filter((line) => line.startsWith("data:"))
    .map((line) => line.slice(5).trimStart());
  if (dataLines.length === 0) return null;
  try {
    const payload: unknown = JSON.parse(dataLines.join("\n"));
    return payload !== null && typeof payload === "object" && !Array.isArray(payload)
      ? (payload as Record<string, unknown>)
      : null;
  } catch {
    return null;
  }
}

/** Split buffered text into complete SSE blocks, returning the remainder. */
export function splitSSEBuffer(buffer: string): { blocks: string[]; rest: string } {
  const blocks = buffer.split(/\r?\n\r?\n/);
  const rest = blocks.pop() ?? "";
  return { blocks, rest };
}

export async function* iterSSEPayloads(
  body: ReadableStream<Uint8Array>,
): AsyncGenerator<Record<string, unknown>> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  try {
    for (;;) {
      const { done, value } = await reader.read();
      buffer += decoder.decode(value, { stream: !done });
      const { blocks, rest } = splitSSEBuffer(buffer);
      buffer = rest;
      for (const block of blocks) {
        const payload = parseSSEBlock(block);
        if (payload !== null) yield payload;
      }
      if (done) break;
    }
    if (buffer.trim()) {
      const payload = parseSSEBlock(buffer);
      if (payload !== null) yield payload;
    }
  } finally {
    reader.releaseLock();
  }
}
